File: root/lang/rust/__demo__/algo-demo/src/list/doubly_linked_list.rs
use std::cell::{Ref, RefCell, RefMut};
use std::rc::Rc;

struct Node<T> {
    data: T,
    next: Option<Rc<RefCell<Node<T>>>>,
    prev: Option<Rc<RefCell<Node<T>>>>,
}

pub struct LinkedList<T> {
    head: Option<Rc<RefCell<Node<T>>>>,
    tail: Option<Rc<RefCell<Node<T>>>>,
}

impl<T> Node<T> {
    fn new(data: T) -> Self {
        Self {
            data,
            next: None,
            prev: None,
        }
    }
}

impl<T> LinkedList<T> {
    pub fn new() -> Self {
        Self {
            head: None,
            tail: None,
        }
    }

    pub fn push_front(&mut self, data: T) {
        let new_head = Rc::new(RefCell::new(Node::new(data)));
        match self.head.take() {
            None => {
                self.tail = Some(new_head.clone());
                self.head = Some(new_head);
            }
            Some(old_head) => {
                old_head.borrow_mut().prev = Some(new_head.clone());
                new_head.borrow_mut().next = Some(old_head);
                self.head = Some(new_head.clone());
            }
        }
    }

    pub fn pop_front(&mut self) -> Option<T> {
        self.head.take().map(|old_head| {
            match old_head.borrow_mut().next.take() {
                None => {
                    self.tail.take();
                }
                Some(new_head) => {
                    new_head.borrow_mut().prev.take();
                    self.head = Some(new_head);
                }
            }
            Rc::try_unwrap(old_head).ok().unwrap().into_inner().data
        })
    }

    pub fn peek_front(&self) -> Option<Ref<T>> {
        self.head
            .as_ref()
            .map(|node| Ref::map(node.borrow(), |node| &node.data))
    }

    pub fn peek_front_mut(&mut self) -> Option<RefMut<T>> {
        self.head
            .as_ref()
            .map(|node| RefMut::map(node.borrow_mut(), |node| &mut node.data))
    }

    pub fn push_back(&mut self, data: T) {
        let new_tail = Rc::new(RefCell::new(Node::new(data)));
        match self.tail.take() {
            None => {
                self.head = Some(new_tail.clone());
                self.tail = Some(new_tail);
            }
            Some(old_tail) => {
                new_tail.borrow_mut().prev = Some(old_tail.clone());
                old_tail.borrow_mut().next = Some(new_tail.clone());
                self.tail = Some(new_tail.clone());
            }
        }
    }

    pub fn pop_back(&mut self) -> Option<T> {
        self.tail.take().map(|old_tail| {
            match old_tail.borrow_mut().prev.take() {
                None => {
                    self.head.take();
                }
                Some(new_tail) => {
                    new_tail.borrow_mut().next.take();
                    self.tail = Some(new_tail);
                }
            };
            Rc::try_unwrap(old_tail).ok().unwrap().into_inner().data
        })
    }

    pub fn peek_back(&self) -> Option<Ref<T>> {
        self.tail
            .as_ref()
            .map(|node| Ref::map(node.borrow(), |node| &node.data))
    }

    pub fn peek_back_mut(&mut self) -> Option<RefMut<T>> {
        self.tail
            .as_ref()
            .map(|node| RefMut::map(node.borrow_mut(), |node| &mut node.data))
    }

    pub fn reverse(&mut self) {
        let mut current = self.head.take();
        self.head = self.tail.take();
        while let Some(current_node) = current {
            let next_node = current_node.borrow().next.clone();
            let mut current_node_ref = current_node.borrow_mut();
            current_node_ref.next = current_node_ref.prev.take();
            if let Some(ref next_node_ref) = next_node {
                next_node_ref.borrow_mut().prev = Some(current_node.clone());
            } else {
                self.tail = Some(current_node.clone());
            }
            current = next_node;
        }
    }
}

impl<T: std::fmt::Display> std::fmt::Display for LinkedList<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut node = self.head.clone();
        while let Some(current) = node {
            write!(f, "{} -> ", current.borrow().data)?;
            node = current.borrow().next.clone();
        }
        write!(f, "None")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_push_front() {
        let mut list = LinkedList::new();
        let total = 5;
        for x in 1..=total {
            list.push_front(x);
            println!("list={list}");
        }
    }

    #[test]
    fn test_pop_front() {
        let mut list = LinkedList::new();
        let total = 5;
        (1..=total).for_each(|x| list.push_front(x));
        for x in 1..=total {
            let front = list.pop_front().unwrap();
            println!(" front={front} list={list}");
        }
    }

    #[test]
    fn test_peek_front() {
        let mut list = LinkedList::new();
        let total = 5;
        (1..=total).for_each(|x| list.push_front(x));
        for x in 1..=total {
            let front = list.peek_front().unwrap().clone();
            println!(" front={front} list={list}");
        }
    }

    #[test]
    fn test_push_back() {
        let mut list = LinkedList::new();
        let total = 5;
        for x in 1..=total {
            list.push_back(x);
            println!("list={list}");
        }
    }

    #[test]
    fn test_pop_back() {
        let mut list = LinkedList::new();
        let total = 5;
        (1..=total).for_each(|x| list.push_back(x));
        for x in 1..=total {
            let front = list.pop_back().unwrap();
            println!(" front={front} list={list}");
        }
    }

    #[test]
    fn test_peek_back() {
        let mut list = LinkedList::new();
        let total = 5;
        (1..=total).for_each(|x| list.push_back(x));
        for x in 1..=total {
            let front = list.peek_back().unwrap();
            println!(" front={front} list={list}");
        }
    }

    #[test]
    fn test_reverse() {
        let mut list = LinkedList::new();
        let total = 5;
        (1..=total).for_each(|x| list.push_back(x));
        println!("list={list}");
        list.reverse();
        println!("list={list}");
    }
}

File: root/lang/javascript/__demo__/algo-demo/src/list/doubly_linked_list.ts
import { type IList } from './index'

class ListNode<T> {
  data: T
  prev: ListNode<T> | undefined
  next: ListNode<T> | undefined

  constructor(data: T) {
    this.data = data
  }
}

export class LinkedList<T> implements IList<T> {
  private head: ListNode<T> | undefined
  private tail: ListNode<T> | undefined
  size: number

  constructor() {
    this.head = undefined
    this.tail = undefined
    this.size = 0
  }

  pushFront(data: T): void {
    const newNode = new ListNode<T>(data)
    if (this.head === undefined) {
      this.head = newNode
      this.tail = newNode
    } else {
      newNode.next = this.head
      this.head.prev = newNode
      this.head = newNode
    }
    this.size++
  }

  pushBack(data: T): void {
    const newNode = new ListNode<T>(data)
    if (this.tail === undefined) {
      this.head = newNode
      this.tail = newNode
    } else {
      newNode.prev = this.tail
      this.tail.next = newNode
      this.tail = newNode
    }
    this.size++
  }

  popFront(): T | undefined {
    if (this.head === undefined) {
      return undefined
    } else {
      const front = this.head.data
      if (this.head === this.tail) {
        this.head = undefined
        this.tail = undefined
      } else {
        this.head = this.head.next
        if (this.head !== undefined) {
          this.head.prev = undefined
        }
      }
      this.size--
      return front
    }
  }

  popBack(): T | undefined {
    if (this.head === undefined) {
      return undefined
    } else {
      const back = this.tail?.data
      if (this.head === this.tail) {
        this.head = undefined
        this.tail = undefined
      } else {
        this.tail = this.tail?.prev
        if (this.tail !== undefined) {
          this.tail.next = undefined
        }
      }
      this.size--
      return back
    }
  }

  peekFront(): T | undefined {
    return this.head?.data
  }

  peekBack(): T | undefined {
    return this.tail?.data
  }

  get(index: number): T {
    this.checkIndex(index)
    let curr = this.head
    for (let i = 0; i < index; i++) {
      curr = curr?.next
    }
    curr = curr as ListNode<T>
    return curr.data
  }

  set(index: number, data: T): void {
    this.checkIndex(index)
    let curr = this.head
    for (let i = 0; i < index; i++) {
      curr = curr?.next
    }
    curr = curr as ListNode<T>
    curr.data = data
  }

  del(index: number): void {
    this.checkIndex(index)
    let curr = this.head
    for (let i = 0; i < index; i++) {
      curr = curr?.next
    }
    curr = curr as ListNode<T>
    const prev = curr.prev
    const next = curr.next
    if (prev === undefined) {
      this.head = next
    } else {
      prev.next = next
    }
    if (next === undefined) {
      this.tail = prev
    } else {
      next.prev = prev
    }
    this.size--
  }

  add(index: number, data: T): void {
    this.checkIndex(index)
    const newNode = new ListNode<T>(data)
    let curr = this.head
    let prev = undefined
    for (let i = 0; i < index; i++) {
      prev = curr
      curr = curr?.next
    }
    curr = curr as ListNode<T>
    curr.prev = newNode
    newNode.prev = prev
    newNode.next = curr
    if (prev === undefined) {
      this.head = newNode
    } else {
      prev.next = newNode
    }
    this.size++
  }

  reverse(): void {
    let curr = this.head
    let prev = undefined
    while (curr !== undefined) {
      const next = curr?.next
      curr.next = prev
      curr.prev = next
      prev = curr
      curr = next
    }
    this.tail = this.head
    this.head = prev
  }

  toString(): string {
    const result = []
    let curr = this.head
    while (curr !== undefined) {
      result.push(curr.data)
      curr = curr.next
    }
    return JSON.stringify(result)
  }

  private checkIndex(index: number): void {
    if (index < 0 || index > this.size) {
      throw new RangeError('Index out of bounds')
    }
  }
}
